Fix manual setup when roomba is on different subnet

jorgror opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix an issue with manual setup of Roomba where it is not discovered automatically. Not my issue, but this describes exactly my setup and I have the same issue (#48180).

My solution does not solve the root cause of the bug, but it avoid it by using an alternative method for getting the name of the Roomba. The root cause of the issue is that name is not available in `async_connect_or_timeout()` in `__init__.py` line 88

A side effect of this fix is that we can also skip one (difficult) step in manual setup.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: 
- This PR is related to issue: #48180
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jorgror committed:
Use _async_discover_roombas to discover roomba
Does not require manual input of BLID
Will also get name, which has a side effect of skipping validate_input() which fail setup of my roomba.
I think the setup issue was discovered in #48180
This does not solve the root cause of the issue which is that name is not available in async_connect_or_timeout() in __init__.py line 88

jorgror committed:
Reduce HOST_ATTEMPTS to 2 as waiting for ~ 60 seconds seems to much

bdraco commented on homeassistant/components/roomba/config_flow.py:
```diff
@@ -24,7 +24,7 @@
 ROOMBA_DISCOVERY_LOCK = "roomba_discovery_lock"
 ALL_ATTEMPTS = 2
-HOST_ATTEMPTS = 6
+HOST_ATTEMPTS = 2
```
Changing this will increase the chance the roomba is missed because it is won't be long enough for the device to see if due to its sleep patterns.

jorgror committed:
Revert "Reduce HOST_ATTEMPTS to 2 as waiting for ~ 60 seconds seems to much"

This reverts commit 6e27bb6dfc38524e513bd09a5cc22042dbf51f83.

jorgror committed:
Fix config_flow tests
Remove BLID as this input is not required anymore
Patch with _mocked_discovery to find device and BLID
Change name of found robot as we use _mocked_discovery to find name (the main reason for this change)

jorgror commented:
> Changing this will increase the chance the roomba is missed because it is won't be long enough for the device to see the discovery request due to its sleep patterns.  

I have reverted this commit as I was not aware of this.

bdraco commented on homeassistant/components/roomba/translations/de.json:
```diff
@@ -31,10 +31,9 @@
             },
             "manual": {
                 "data": {
```
No need to change anything but strings.json and `en.json` as the rest will be overwritten anyways by lokalize

https://developers.home-assistant.io/docs/translations/

jorgror replied:
I have reverted these changes. Thanks for the tip.

jorgror committed:
Updated strings to remove BLID field from manual setup.
Updated translations to remove BLID description.

jorgror committed:
Mocked discovery to speed up test

bdraco commented on homeassistant/components/roomba/config_flow.py:
```diff
@@ -175,17 +175,21 @@ async def async_step_manual(self, user_input=None):
-        self.blid = user_input[CONF_BLID].upper()
+
+        devices = await _async_discover_roombas(self.hass, self.host)
+        if devices:
+            self.blid = devices[0].blid
+            self.name = devices[0].robot_name
+        else:
+            return self.async_abort(reason="cannot_connect")
```
```suggestion
        if not devices:
            return self.async_abort(reason="cannot_connect")
        self.blid = devices[0].blid
        self.name = devices[0].robot_name
            
```

jorgror committed:
Update homeassistant/components/roomba/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco reviewed (commented):
Tested and working 👍 

I think it just needs `black` and should be good to go.

frenck committed:
Ran black

frenck commented:
Ran black.

bdraco reviewed (approved):
LGTM 👍
